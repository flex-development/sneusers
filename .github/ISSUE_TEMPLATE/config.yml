# ISSUE TEMPLATE CHOOSER
# https://docs.github.com/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#configuring-the-template-chooser

---
blank_issues_enabled: true
contact_links:
  - name: Discuss an idea
    url: https://github.com/flex-development/sneusers/discussions/new?category=ideas
    about: Discuss an idea before submitting a feature request
  - name: Q&A
    url: https://github.com/flex-development/sneusers/discussions/new?category=q-a
    about: Ask and answer questions
