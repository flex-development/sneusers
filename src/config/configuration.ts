import { ObjectPlain } from '@flex-development/tutils'
import { NodeEnv } from '@sneusers/enums'
import { EnvironmentVariables } from '@sneusers/models'
import { instanceToPlain } from 'class-transformer'
import { validateSync, ValidationError } from 'class-validator'

/**
 * @file Configuration - Environment Variables
 * @module sneusers/config/configuration
 * @see https://docs.nestjs.com/techniques/configuration
 */

/** @property {string[]} ENV_FILE_PATH - Custom environment files */
const ENV_FILE_PATH = [`${process.cwd()}/.env.local`, `${process.cwd()}/.env`]

/**
 * Validates environment variables.
 *
 * @param {ObjectPlain} config - Object containing environment variables
 * @return {EnvironmentVariables} **Validated** environment variables
 * @throws {ValidationError[]}
 */
const validate = ({
  CACHE_MAX = 100,
  CACHE_TTL = 5,
  DB_AUTO_LOAD_MODELS = 'true',
  DB_HOST = 'localhost',
  DB_NAME,
  DB_PASSWORD,
  DB_PORT = 3306,
  DB_USERNAME,
  EMAIL_CLIENT,
  EMAIL_HOST = 'smtp.gmail.com',
  EMAIL_PORT = 465,
  EMAIL_PRIVATE_KEY,
  EMAIL_SEND_AS,
  EMAIL_USER,
  HOST,
  HOSTNAME = 'localhost',
  JWT_EXP_ACCESS = 900,
  JWT_EXP_REFRESH = 86_400,
  JWT_EXP_VERIFY = 86_400,
  JWT_SECRET_ACCESS,
  JWT_SECRET_REFRESH,
  JWT_SECRET_VERIFY,
  NODE_ENV = NodeEnv.DEV,
  PORT = 8080,
  REDIS_HOST = 'redis',
  REDIS_PORT = 6379,
  THROTTLE_LIMIT = 10,
  THROTTLE_TTL = 60
}: ObjectPlain): EnvironmentVariables => {
  const env = new EnvironmentVariables()

  // Set Node environment
  env.NODE_ENV = NODE_ENV

  // Check if running in local production mode or deployed production mode
  env.PROD_LOCAL = env.NODE_ENV === NodeEnv.PROD && HOSTNAME === 'localhost'
  env.PROD = env.NODE_ENV === NodeEnv.PROD && !env.PROD_LOCAL

  // Set hostname and port to run application on
  env.HOSTNAME = HOSTNAME
  env.PORT = Number.parseInt(PORT.toString())

  // Assign remaining environment variables
  env.CACHE_MAX = Number.parseInt(CACHE_MAX.toString())
  env.CACHE_TTL = Number.parseInt(CACHE_TTL.toString())
  env.DB_AUTO_LOAD_MODELS = JSON.parse(DB_AUTO_LOAD_MODELS)
  env.DB_HOST = DB_HOST
  env.DB_NAME = DB_NAME
  env.DB_PASSWORD = DB_PASSWORD
  env.DB_PORT = Number.parseInt(DB_PORT.toString(), 10)
  env.DB_USERNAME = DB_USERNAME
  env.DEV = NODE_ENV === NodeEnv.DEV
  env.EMAIL_CLIENT = EMAIL_CLIENT
  env.EMAIL_HOST = EMAIL_HOST
  env.EMAIL_PORT = Number.parseInt(EMAIL_PORT.toString(), 10)
  env.EMAIL_PRIVATE_KEY = EMAIL_PRIVATE_KEY!.replace(/\\n/g, '\n')
  env.EMAIL_SEND_AS = EMAIL_SEND_AS
  env.EMAIL_USER = EMAIL_USER
  env.HOST = HOST || `http://${env.HOSTNAME}:${env.PORT}`
  env.JWT_EXP_ACCESS = Number.parseInt(JWT_EXP_ACCESS.toString(), 10)
  env.JWT_EXP_REFRESH = Number.parseInt(JWT_EXP_REFRESH.toString(), 10)
  env.JWT_EXP_VERIFY = Number.parseInt(JWT_EXP_VERIFY.toString(), 10)
  env.JWT_SECRET_ACCESS = (env.PROD && JWT_SECRET_ACCESS) || 'JWT_SECRET'
  env.JWT_SECRET_REFRESH = (env.PROD && JWT_SECRET_REFRESH) || 'JWT_SECRET'
  env.JWT_SECRET_VERIFY = (env.PROD && JWT_SECRET_VERIFY) || 'JWT_SECRET'
  env.REDIS_HOST = REDIS_HOST
  env.REDIS_PORT = Number.parseInt(REDIS_PORT.toString(), 10)
  env.TEST = NODE_ENV === NodeEnv.TEST
  env.THROTTLE_LIMIT = Number.parseInt(THROTTLE_LIMIT.toString(), 10)
  env.THROTTLE_TTL = Number.parseInt(THROTTLE_TTL.toString(), 10)

  // Validate environment variables
  const errors: ValidationError[] = validateSync(env, {
    enableDebugMessages: true,
    forbidUnknownValues: true,
    stopAtFirstError: false,
    validationError: { target: false }
  })

  // Throw errors if found
  if (errors.length > 0) throw errors

  // Return validated environment variables as plain object
  return instanceToPlain(env) as EnvironmentVariables
}

/**
 * Returns an object containing the application's environment variables.
 *
 * @return {EnvironmentVariables} **Validated** environment variables
 */
const configuration = (): EnvironmentVariables => {
  return validate({
    CACHE_MAX: process.env.CACHE_MAX,
    CACHE_TTL: process.env.CACHE_TTL,
    DB_AUTO_LOAD_MODELS: process.env.DB_AUTO_LOAD_MODELS,
    DB_HOST: process.env.DB_HOST,
    DB_NAME: process.env.DB_NAME,
    DB_PASSWORD: process.env.DB_PASSWORD,
    DB_PORT: process.env.DB_PORT,
    DB_USERNAME: process.env.DB_USERNAME,
    EMAIL_CLIENT: process.env.EMAIL_CLIENT,
    EMAIL_HOST: process.env.EMAIL_HOST,
    EMAIL_PORT: process.env.EMAIL_PORT,
    EMAIL_PRIVATE_KEY: process.env.EMAIL_PRIVATE_KEY,
    EMAIL_SEND_AS: process.env.EMAIL_SEND_AS,
    EMAIL_USER: process.env.EMAIL_USER,
    HOST: process.env.HOST,
    HOSTNAME: process.env.HOSTNAME,
    JWT_EXP_ACCESS: process.env.JWT_EXP_ACCESS,
    JWT_EXP_REFRESH: process.env.JWT_EXP_REFRESH,
    JWT_EXP_VERIFY: process.env.JWT_EXP_VERIFY,
    JWT_SECRET_ACCESS: process.env.JWT_SECRET_ACCESS,
    JWT_SECRET_REFRESH: process.env.JWT_SECRET_REFRESH,
    JWT_SECRET_VERIFY: process.env.JWT_SECRET_VERIFY,
    NODE_ENV: process.env.NODE_ENV,
    PORT: process.env.PORT,
    REDIS_HOST: process.env.REDIS_HOST,
    REDIS_PORT: process.env.REDIS_PORT,
    THROTTLE_LIMIT: process.env.THROTTLE_LIMIT,
    THROTTLE_TTL: process.env.THROTTLE_TTL
  })
}

/** @property {EnvironmentVariables} ENV - Application environment variables */
const ENV: EnvironmentVariables = configuration()

export { ENV, ENV_FILE_PATH, configuration as default, validate }
