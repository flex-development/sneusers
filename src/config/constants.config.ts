import type { NestApplicationOptions as NestAppOptions } from '@nestjs/common'
import pkg from 'read-pkg'

/**
 * @file Configuration - Constant Values
 * @module sneusers/config/constants
 */

/**
 * @property {string} CURRENT_TIMESTAMP - Stringified `strftime` call
 * @see https://www.w3resource.com/sqlite/sqlite-strftime.php
 */
export const CURRENT_TIMESTAMP: string = "strftime('%s','now')"

/**
 * @property {string} DEBUG_NAMESPACE - Name of `debug` namespace
 * @see https://github.com/visionmedia/debug
 */
export const DEBUG_NAMESPACE: string = 'sneusers'

/** @property {Package} PACKAGE - `package.json` data */
export const PACKAGE: Package = pkg.sync({ normalize: false }) as Package

/** @property {NestAppOptions} NEST_APP_OPTIONS - NestJS application options */
export const NEST_APP_OPTIONS: NestAppOptions = {
  cors: true,
  logger: console
}
