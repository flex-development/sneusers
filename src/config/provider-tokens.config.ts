import { forwardRef, ForwardReference as ForwardRef } from '@nestjs/common'

/**
 * @file Configuration - Provider Tokens
 * @module sneusers/config/provider-tokens
 */

/** @property {ForwardRef<string>} AXIOS_INSTANCE - `AxiosInstance` */
export const AXIOS_INSTANCE: ForwardRef<string> = forwardRef(() => {
  return 'AXIOS_INSTANCE_TOKEN'
})
