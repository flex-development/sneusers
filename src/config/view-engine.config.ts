import { create } from 'express-handlebars'
import path from 'path'

/**
 * @file Configuration - viewEngine
 * @module sneusers/config/viewEngine
 */

export default create({
  /** @see https://github.com/ericf/express-handlebars/pull/249 */
  defaultLayout: null as unknown as string,
  extname: '.hbs',
  partialsDir: path.join(process.cwd(), 'views', 'partials')
})
