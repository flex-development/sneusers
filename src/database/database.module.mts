/**
 * @file DatabaseModule
 * @module sneusers/database/DatabaseModule
 */

import Mapper from '#database/providers/base.mapper'
import type { Class } from '@flex-development/tutils'
import {
  Global,
  Module,
  type DynamicModule,
  type InjectionToken,
  type Provider
} from '@nestjs/common'

/**
 * Database module.
 *
 * @class
 */
@Global()
@Module({})
class DatabaseModule {
  /**
   * Create a data mapper class.
   *
   * @public
   * @static
   *
   * @param {Class<any>} Entity
   *  Database entity model
   * @return {Class<Mapper>}
   *  Data mapper class
   */
  public static Mapper(Entity: Class<any>): Class<Mapper> {
    /**
     * Entity data mapper.
     *
     * @class
     * @extends {Mapper}
     */
    return class EntityMapper extends Mapper {
      /**
       * @public
       * @static
       * @override
       * @member {string} name
       */
      public static override name: string = Entity.name + 'Mapper'

      /**
       * @protected
       * @instance
       * @member {Class<any>} Entity
       */
      protected Entity: Class<any> = Entity

      /**
       * Create a new data mapper.
       */
      constructor() {
        super()
        Object.assign(this, { name: EntityMapper.name })
      }
    }
  }

  /**
   * Register database entities.
   *
   * @public
   * @static
   *
   * @param {Class<any>[]} entities
   *  List of database entities to register
   * @return {DynamicModule}
   *  Dynamic feature module
   */
  public static forFeature(...entities: Class<any>[]): DynamicModule {
    /**
     * List of injection tokens representing exported providers.
     *
     * @const {InjectionToken[]} exports
     */
    const exports: InjectionToken[] = []

    /**
     * List of feature module providers.
     *
     * @const {Provider[]} providers
     */
    const providers: Provider[] = []

    // provide data mappers.
    for (const Entity of entities) {
      /**
       * Data mapper class.
       *
       * @const {Class<Mapper>} EntityMapper
       */
      const EntityMapper: Class<Mapper> = DatabaseModule.Mapper(Entity)

      providers.push({ provide: EntityMapper.name, useClass: EntityMapper })
      exports.push(EntityMapper.name)
    }

    return { exports, global: false, module: DatabaseModule, providers }
  }
}

export default DatabaseModule
