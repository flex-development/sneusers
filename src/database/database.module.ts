/**
 * @file Database - DatabaseModule
 * @module sneusers/database/DatabaseModule
 */

import type { IConfig } from '#src/interfaces'
import * as pathe from '@flex-development/pathe'
import { NodeEnv } from '@flex-development/tutils'
import {
  ReflectMetadataProvider,
  UnderscoreNamingStrategy
} from '@mikro-orm/core'
import { MikroOrmModule, type MikroOrmModuleOptions } from '@mikro-orm/nestjs'
import { Global, Module } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import { template } from 'radash'

/**
 * Database module.
 *
 * Database driver: MongoDB
 *
 * @see https://docs.nestjs.com/recipes/mikroorm
 * @see https://mikro-orm.io
 *
 * @class
 */
@Global()
@Module({
  imports: [
    MikroOrmModule.forRootAsync({
      inject: [ConfigService],
      useFactory: (
        config: ConfigService<IConfig, true>
      ): Omit<MikroOrmModuleOptions, 'contextName'> => {
        /**
         * Test environment check.
         *
         * @const {boolean} TEST
         */
        const TEST: boolean = config.get<NodeEnv>('NODE_ENV') === NodeEnv.TEST

        return {
          allowGlobalContext: false,
          autoLoadEntities: true,
          clientUrl: template('mongodb://{{0}}:{{1}}@{{2}}:{{3}}/', {
            0: config.get<string>('DB_USERNAME'),
            1: config.get<string>('DB_PASSWORD'),
            2: config.get<string>('DB_HOSTNAME'),
            3: config.get<number>('DB_PORT')
          }),
          connect: !TEST,
          dbName: config.get<string>('DB_NAME'),
          debug: !TEST,
          discovery: {
            alwaysAnalyseProperties: false,
            requireEntitiesArray: true,
            warnWhenNoEntities: false
          },
          ensureDatabase: false,
          ensureIndexes: false,
          forceEntityConstructor: true,
          forceUndefined: false,
          forceUtcTimezone: true,
          metadataProvider: ReflectMetadataProvider,
          migrations: {
            allOrNothing: true,
            disableForeignKeys: true,
            dropTables: true,
            emit: 'ts',
            glob: '!(*.d).ts',
            path: pathe.resolve('dist/database/migrations'),
            pathTs: pathe.resolve('src/database/migrations'),
            safe: false,
            snapshot: true,
            tableName: 'migrations',
            transactional: true
          },
          namingStrategy: UnderscoreNamingStrategy,
          password: config.get<string>('DB_PASSWORD'),
          persistOnCreate: false,
          seeder: {
            path: pathe.resolve('dist/database/seeders'),
            pathTs: pathe.resolve('src/database/seeders')
          },
          strict: true,
          tsNode: true,
          type: 'mongo',
          user: config.get<string>('DB_USERNAME'),
          validate: true,
          validateRequired: true
        }
      }
    })
  ]
})
class DatabaseModule {}

export default DatabaseModule
