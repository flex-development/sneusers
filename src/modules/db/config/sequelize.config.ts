import sequelize from '@fixtures/sequelize.fixture'
import { ENV } from '@sneusers/config/configuration'
import { DatabaseTable } from '../enums'

/**
 * @file DatabaseModule Configuration - Sequelize
 * @module sneusers/modules/db/config/sequelize
 */

export = {
  [ENV.APP_ENV]: {
    ...sequelize.options,
    migrationStorageTableName: DatabaseTable.STORAGE_MIGRATIONS,
    seederStorage: 'sequelize',
    seederStorageTableName: DatabaseTable.STORAGE_SEEDERS
  }
}
