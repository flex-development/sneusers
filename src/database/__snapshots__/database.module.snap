// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:database/DatabaseModule > .forFeature > should return dynamic feature module 1`] = `
{
  "exports": [
    "EntityMapper",
  ],
  "global": false,
  "module": [Function DatabaseModule],
  "providers": [
    {
      "provide": "EntityMapper",
      "useClass": [Function EntityMapper],
    },
  ],
}
`;
